src/rust/cryptography-x509/src/common.rs

```diff
@@ -301,9 +301,117 @@ impl<'a> asn1::SimpleAsn1Writable for UnvalidatedVisibleString<'a> {
     }
 }
 
+/// A `DNSName` is an `asn1::IA5String` with additional invariant preservations
+/// per [RFC 5280 4.2.1.6], which in turn uses the preferred name syntax defined
+/// in [RFC 1034 3.5] and amended in [RFC 1123 2.1].
+///
+/// Non-ASCII domain names (i.e., internationalized names) must be pre-encoded;
+/// comparisons are case-insensitive.
+///
+/// [RFC 5280 4.2.1.6]: https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.6
+/// [RFC 1034 3.5]: https://datatracker.ietf.org/doc/html/rfc1034#section-3.5
+/// [RFC 1123 2.1]: https://datatracker.ietf.org/doc/html/rfc1123#section-2.1
+///
+/// ```rust
+/// # use cryptography_x509::common::DNSName;
+/// assert_eq!(DNSName::new("foo.com").unwrap(), DNSName::new("FOO.com").unwrap());
+/// ```
+#[derive(Debug)]
+pub struct DNSName<'a>(asn1::IA5String<'a>);
+
+impl<'a> DNSName<'a> {
+    pub fn new(value: &'a str) -> Option<Self> {
+        // Domains cannot be empty; cannot contain whitespace; must (practically)
+        // be less than 253 characters (255 in RFC 1034's octet encoding).
+        if value.is_empty() || value.chars().any(char::is_whitespace) || value.len() > 253 {
```

Comment:
Is the whitespace check not covered by the loop below? (Trying to avoid iterating the string 2x)

Reply:
Yep, it is -- thanks for catching that!